File: docs/postman/rootAuthSigner.mjs
import http from "node:http";
import { spawnSync } from "node:child_process";
import { createHash } from "node:crypto";
import { chmodSync, mkdtempSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

const homeDirectory = process.env.HOME ?? process.env.USERPROFILE ?? "";
const privateKeyPath =
  process.env.ROOT_AUTH_SIGNER_PRIVATE_KEY_PATH ?? join(homeDirectory, ".ssh", "id_ed25519");
const bindHost = process.env.ROOT_AUTH_SIGNER_HOST ?? "127.0.0.1";
const bindPort = Number(process.env.ROOT_AUTH_SIGNER_PORT ?? "8787");
const signingNamespace = "kanbien-platform";
const tempRootsToCleanup = [];

function registerTempRoot(tempRoot) {
  tempRootsToCleanup.push(tempRoot);
  return tempRoot;
}

function cleanupTempRoots() {
  for (const tempRoot of tempRootsToCleanup.splice(0)) {
    rmSync(tempRoot, { recursive: true, force: true });
  }
}

process.on("exit", cleanupTempRoots);
process.on("SIGINT", () => {
  cleanupTempRoots();
  process.exit(130);
});
process.on("SIGTERM", () => {
  cleanupTempRoots();
  process.exit(143);
});

function runSshKeygen(args, options = {}) {
  const result = spawnSync("ssh-keygen", args, {
    stdio: ["pipe", "pipe", "pipe"],
    windowsHide: true,
    ...options,
  });

  if (result.error) {
    if (result.error.code === "ENOENT") {
      throw new Error("OpenSSH ssh-keygen is required before the signer helper can run.");
    }

    throw result.error;
  }

  if (result.status !== 0) {
    const stderr = result.stderr ? String(result.stderr).trim() : "";
    const stdout = result.stdout ? String(result.stdout).trim() : "";
    throw new Error(stderr || stdout || "ssh-keygen failed.");
  }

  return {
    stdout: result.stdout ? String(result.stdout) : "",
    stderr: result.stderr ? String(result.stderr) : "",
  };
}

function loadPublicKeyOpenSsh(keyPath) {
  if (keyPath.endsWith(".pub")) {
    throw new Error("ROOT_AUTH_SIGNER_PRIVATE_KEY_PATH must point to a private key file.");
  }

  return runSshKeygen(["-y", "-f", keyPath]).stdout.trim();
}

function preparePrivateKeyPath(keyPath) {
  if (!keyPath.startsWith("/mnt/")) {
    return keyPath;
  }

  const tempRoot = registerTempRoot(mkdtempSync(join(tmpdir(), "kanbien-postman-key-")));
  const stagedKeyPath = join(tempRoot, "signing-key");
  writeFileSync(stagedKeyPath, readFileSync(keyPath));
  chmodSync(stagedKeyPath, 0o600);
  return stagedKeyPath;
}

function toSshEd25519Fingerprint(publicKeyOpenSsh) {
  const parts = publicKeyOpenSsh.trim().split(/\s+/);

  if (parts.length < 2 || parts[0] !== "ssh-ed25519") {
    throw new Error("Only ssh-ed25519 keys are supported by this signer.");
  }

  const decoded = Buffer.from(parts[1], "base64");
  return `SHA256:${createHash("sha256").update(decoded).digest("base64").replace(/=+$/, "")}`;
}

function signChallengeText(challengeText, keyPath) {
  const tempRoot = mkdtempSync(join(tmpdir(), "kanbien-postman-signer-"));
  const challengePath = join(tempRoot, "challenge.txt");
  const signaturePath = `${challengePath}.sig`;

  try {
    writeFileSync(challengePath, challengeText, "utf8");
    runSshKeygen(["-Y", "sign", "-f", keyPath, "-n", signingNamespace, challengePath]);
    return readFileSync(signaturePath, "utf8").trim();
  } finally {
    rmSync(tempRoot, { recursive: true, force: true });
  }
}

const effectivePrivateKeyPath = preparePrivateKeyPath(privateKeyPath);
const publicKeyOpenSsh = loadPublicKeyOpenSsh(effectivePrivateKeyPath);
const publicKeyFingerprint = toSshEd25519Fingerprint(publicKeyOpenSsh);

const server = http.createServer((req, res) => {
  const requestPath = req.url ? new URL(req.url, `http://${bindHost}:${bindPort}`).pathname : "";

  if (req.method !== "POST" || !["/sign", "/sign/"].includes(requestPath)) {
    res.writeHead(404, { "Content-Type": "application/json" });
    res.end(JSON.stringify({ code: "NOT_FOUND", message: "Not found" }));
    return;
  }

  let body = "";
  req.on("data", (chunk) => {
    body += chunk;
  });

  req.on("end", () => {
    try {
      const parsed = JSON.parse(body);
      const challengeText = parsed.challengeText;

      if (typeof challengeText !== "string" || challengeText.trim().length === 0) {
        res.writeHead(400, { "Content-Type": "application/json" });
        res.end(
          JSON.stringify({
            code: "INVALID_REQUEST",
            message: "challengeText is required",
          }),
        );
        return;
      }

      const signature = signChallengeText(challengeText, effectivePrivateKeyPath);

      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(
        JSON.stringify({
          signature,
          publicKeyFingerprint,
        }),
      );
    } catch (error) {
      res.writeHead(500, { "Content-Type": "application/json" });
      res.end(
        JSON.stringify({
          code: "SIGNING_FAILED",
          message: error instanceof Error ? error.message : "Unknown signing failure",
        }),
      );
    }
  });
});

server.listen(bindPort, bindHost, () => {
  console.log(`Root auth signer listening on http://${bindHost}:${bindPort}`);
  console.log(`Using private key: ${privateKeyPath}`);
  if (effectivePrivateKeyPath !== privateKeyPath) {
    console.log(`Staged private key for OpenSSH permission checks: ${effectivePrivateKeyPath}`);
  }
  console.log(`Fingerprint: ${publicKeyFingerprint}`);
  console.log("Using system ssh-keygen for signing.");
});

File: docs/postman/rootAuthSigner.py
import base64
import json
import os
from hashlib import sha256
from http.server import BaseHTTPRequestHandler, HTTPServer
from pathlib import Path

from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey

DEFAULT_PRIVATE_KEY_PATH = str(Path.home() / ".ssh" / "id_ed25519")
PRIVATE_KEY_PATH = os.environ.get("ROOT_AUTH_SIGNER_PRIVATE_KEY_PATH", DEFAULT_PRIVATE_KEY_PATH)
HOST = os.environ.get("ROOT_AUTH_SIGNER_HOST", "127.0.0.1")
PORT = int(os.environ.get("ROOT_AUTH_SIGNER_PORT", "8787"))


with open(PRIVATE_KEY_PATH, "rb") as key_file:
    private_key = serialization.load_ssh_private_key(key_file.read(), password=None)

if not isinstance(private_key, Ed25519PrivateKey):
    raise RuntimeError("Only Ed25519 private keys are supported by this signer")

public_key = private_key.public_key()
public_key_bytes = public_key.public_bytes(
    encoding=serialization.Encoding.Raw,
    format=serialization.PublicFormat.Raw,
)


def ssh_string(raw: bytes) -> bytes:
    return len(raw).to_bytes(4, "big") + raw


def build_fingerprint() -> str:
    blob = ssh_string(b"ssh-ed25519") + ssh_string(public_key_bytes)
    digest = sha256(blob).digest()
    return "SHA256:" + base64.b64encode(digest).decode("ascii").rstrip("=")


PUBLIC_KEY_FINGERPRINT = build_fingerprint()


class SignHandler(BaseHTTPRequestHandler):
    def _send_json(self, status: int, payload: dict) -> None:
        body = json.dumps(payload).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def do_POST(self) -> None:
        if self.path != "/sign":
            self._send_json(404, {"code": "NOT_FOUND", "message": "Not found"})
            return

        length = int(self.headers.get("Content-Length", "0"))
        raw_body = self.rfile.read(length)

        try:
            payload = json.loads(raw_body.decode("utf-8"))
            challenge_text = payload.get("challengeText")
            if not isinstance(challenge_text, str) or not challenge_text:
                self._send_json(
                    400,
                    {"code": "INVALID_REQUEST", "message": "challengeText is required"},
                )
                return

            signature = private_key.sign(challenge_text.encode("utf-8"))
            self._send_json(
                200,
                {
                    "signature": base64.b64encode(signature).decode("ascii"),
                    "publicKeyFingerprint": PUBLIC_KEY_FINGERPRINT,
                },
            )
        except Exception as error:
            self._send_json(500, {"code": "SIGNING_FAILED", "message": str(error)})

    def log_message(self, _format: str, *_args) -> None:
        return


if __name__ == "__main__":
    server = HTTPServer((HOST, PORT), SignHandler)
    print(f"Root auth signer listening on http://{HOST}:{PORT}")
    print(f"Using private key: {PRIVATE_KEY_PATH}")
    print(f"Fingerprint: {PUBLIC_KEY_FINGERPRINT}")
    print("Override ROOT_AUTH_SIGNER_PRIVATE_KEY_PATH if you want to use a different key.")
    server.serve_forever()
